spec/compromise-syntax.md

```diff
@@ -0,0 +1,238 @@
+# MF2.0 compromise syntax
+
+# Intro
+
+This syntax builds on the one from https://github.com/unicode-org/message-format-wg/pull/230
+but modified to address
+[@markusicu’s comments there](https://github.com/unicode-org/message-format-wg/pull/230#issuecomment-1116903103).
+
+# Basic syntax
+
+Messages need to delineate between literal text, placeholders, and other “code”.
+We should start in “code mode” and always enclose “patterns” (text+placeholders) in curly braces.
+```
+{Hello world!}
+{Hello {$name}!}
+```
+
+This is unusual for formatting syntaxes, but useful.
+We anyway need to support selecting from among multiple patterns,
+and delimiting the patterns makes it unambiguous
+what white space is part of the pattern vs. serves as delimiters of “code” tokens.
+For consistency, we should always enclose a pattern,
+even if the message consists only of that pattern.
+That also helps with embedding messages in various resource file formats,
+because they can freely trim surrounding white space without
+requiring escapes when a message pattern wants to start or end with spaces.
+
+By contrast, consider the experience with the existing ICU MessageFormat syntax
+which does start in “text mode”.
+ICU MessageFormat has pioneered the selection among multiple patterns based on run-time arguments.
+It represents selection using complex placeholders,
+which has the side effect of allowing literal text and other placeholders
+before and after the top-level selection placeholder.
+However, for reliable translations,
+there should be no translatable contents before or after the selection placeholder;
+instead, each selectable pattern should form one complete “translation unit”.
+Because the existing ICU MessageFormat starts in “text mode”,
+even though it looks like there is no extraneous text,
+spurious white space creeps in from developers’ line breaking of long message strings.
+The remedy is to always use syntax to indicate the start of translatable contents.
+
+We use curly braces to delimit patterns because
+`{}` are the paired ASCII punctuation characters least commonly used in normal text.
+For the same reason, we also use them for embedding placeholders in patterns.
+
+Literal text can use any characters except for curly braces,
+and except for the backslash, which we use as usual for escaping.
+That is, the only special characters inside a pattern are `{}\`.
+The only allowed escape sequences are `\{`, `\}`, and `\\`.
+It is an error if `\` is followed by any other character.
+
+The message syntax does not use `'` or `"`,
+so that it is easy to hard-code message strings in programming language source code.
+
+# Placeholders
+
+Formatting a message replaces placeholders with values based on run-time arguments or special functions.
+We also allow for value literals specified inside the placeholder,
+instead of using an argument name;
+and we also allow for invoking functions without using argument names or value literals.
+```
+{$name}
```

Comment:
I guess I'm discovering a little reluctance around using `$` as the variable identifier, mainly based on "I have tons of strings with place holders like `{someVar}` that need to be `{$someVar}`. It also means that I can't just take my arg map--I need to decorate the variable names with a `$` before I can use it. Since function and format names are decorated with a `:` and literals are delimited with `<>`, do we need the `$`?

Reply:
This compromise syntax builds on what the committee has done. AFAICT the dollar prefix seemed part of what consensus was able to form. I am personally not particularly wedded to it.

For a parser, it would be slightly easier to look for one of very few special characters. If argument names didn't have a prefix, then a parser would have to look for any identifier-start character. Given that it has to anyway do so immediately after a prefix character, it would not really add significant complication. It just comes down to what we think developers reading and writing message strings will find helpful or confusing.